std/encoding.d

```diff
@@ -3380,3 +3380,169 @@ version(unittest)
+    BOMSeq(BOM.utf7, to!(ubyte[])([0x2B, 0x2F, 0x76, 0x38, 0x2D])),
+    BOMSeq(BOM.utf7, to!(ubyte[])([0x2B, 0x2F, 0x76, 0x38])),
+    BOMSeq(BOM.utf1, to!(ubyte[])([0xF7, 0x64, 0x4C])),
+    BOMSeq(BOM.utfebcdic, to!(ubyte[])([0xDD, 0x73, 0x66, 0x73])),
+    BOMSeq(BOM.scsu, to!(ubyte[])([0x0E, 0xFE, 0xFF])),
+    BOMSeq(BOM.bocu1, to!(ubyte[])([0xFB, 0xEE, 0x28])),
+    BOMSeq(BOM.gb18030, to!(ubyte[])([0x84, 0x31, 0x95, 0x33])),
+    BOMSeq(BOM.utf8, to!(ubyte[])([0xEF, 0xBB, 0xBF])),
+    BOMSeq(BOM.utf16be, to!(ubyte[])([0xFE, 0xFF])),
+    BOMSeq(BOM.utf16le, to!(ubyte[])([0xFF, 0xFE]))
+];
+
+/** Returns a $(D BOMSeq) for a given $(D input).
+If no $(D BOM) is present the $(D BOMSeq) for $(D BOM.none) is
+returned. The $(D BOM) sequence at the beginning of the range will
+not be comsumed from the passed range. If you pass a reference type
+range make sure that $(D save) creates a deep copy.
+
+Params:
+    input = The sequence to check for the $(D BOM)
+
+Returns:
+    the found $(D BOMSeq) corresponding to the passed $(D input).
+*/
+immutable(BOMSeq) getBOM(Range)(Range input)
+        if(isForwardRange!Range && is(Unqual!(ElementType!(Range)) == ubyte))
+{
+    import std.algorithm.searching : startsWith;
+    foreach (it; bomTable[1 .. $])
+    {
+        if (startsWith(input.save, it.sequence))
+        {
+            return it;
+        }
+    }
+
+    return bomTable[0];
+}
+
+///
+unittest
+{
+    import std.format : format;
+
+    // This creates a dstring "Hello World" with a BOM
+    auto ts = (cast(dstring)to!(ubyte[])([0xFF, 0xFE, 0x00, 0x00])) ~
+        "Hello World"d;
+
+    auto entry = getBOM(cast(ubyte[])ts);
+    assert(entry.schema == BOM.utf32le);
+}
+
+unittest
+{
+    import std.format : format;
+
+    foreach (idx, it; bomTable)
+    {
+        auto s = it[1] ~ cast(ubyte[])"hello world";
+        auto i = getBOM(s);
+        assert(i[0] == bomTable[idx][0]);
+
+        if (idx < 4 || idx > 7) // get around the multiple utf7 bom's
+        {
+            assert(i[0] == BOM.init + idx);
+            assert(i[1] == it[1]);
+        }
+    }
+}
+
+unittest
+{
+    struct BOMInputRange
+    {
+        immutable(ubyte[]) arr;
+        size_t idx = 0;
+
+        @property ubyte front()
+        {
+            return this.arr[idx];
+        }
+
+        @property bool empty()
+        {
+            return this.idx == this.arr.length;
+        }
+
+        void popFront()
+        {
+            ++this.idx;
+        }
+
+        @property typeof(this) save()
+        {
+            return this;
+        }
+    }
+
+    static assert( isInputRange!BOMInputRange);
+    static assert(!isArray!BOMInputRange);
+
+    ubyte[] dummyEnd = [0,0,0,0];
+
+    foreach (idx, it; bomTable[1 .. $])
+    {
+        {
+            auto ir = BOMInputRange(it.sequence);
+
+            auto b = getBOM(ir);
+            assert(b.schema == it.schema);
+            assert(ir.arr == it.sequence);
```

Comment:
Something is fallacious with this unittest. It's basically saying that the output of `getBOM` should be the same as what the current implementation of `getBOM` does. How does that prove anything??  Shouldn't we be handing literal byte sequences to `getBOM` and verifying that it returns the correct value instead? E.g.,
````
auto data1 = BOMInputRange(cast(ubyte[])[ 0x00, ... /* insert test data here */]);
auto b = getBOM(data1);
assert(b.schema == /* whatever the byte sequence was crafted to be */);
````
Otherwise the unittest is saying the equivalent of `x == x` which proves nothing at all.